pkg/server/test/write.go

```diff
@@ -75,278 +78,183 @@ var writeCommandTests = []writeCommandTest{
 		// state
 		model: &openfgapb.AuthorizationModel{
 			Id:            ulid.Make().String(),
-			SchemaVersion: typesystem.SchemaVersion1_0,
-			TypeDefinitions: []*openfgapb.TypeDefinition{
-				{
-					Type: "repository",
-					Relations: map[string]*openfgapb.Userset{
-						"writer": {Userset: &openfgapb.Userset_This{}},
-						"owner":  {Userset: &openfgapb.Userset_This{}},
-						"viewer": {
-							Userset: &openfgapb.Userset_Union{
-								Union: &openfgapb.Usersets{
-									Child: []*openfgapb.Userset{
-										{Userset: &openfgapb.Userset_ComputedUserset{
-											ComputedUserset: &openfgapb.ObjectRelation{
-												Object:   "",
-												Relation: "writer",
-											},
-										}},
-										{Userset: &openfgapb.Userset_ComputedUserset{
-											ComputedUserset: &openfgapb.ObjectRelation{
-												Object:   "",
-												Relation: "owner",
-											},
-										}},
-									},
-								},
-							},
-						},
-					},
-				},
-			},
+			SchemaVersion: typesystem.SchemaVersion1_1,
+			TypeDefinitions: parser.MustParse(`
+			type user
+
+			type repo
+			  relations
+			    define writer: [user] as self
+				define owner: [user] as self
+				define viewer as writer or owner
+			`),
 		},
 		// input
 		request: &openfgapb.WriteRequest{
 			Writes: &openfgapb.TupleKeys{TupleKeys: []*openfgapb.TupleKey{{
-				Object:   "repository:openfga/openfga",
+				Object:   "repo:openfga/openfga",
 				Relation: "viewer",
-				User:     "github|alice@openfga.com",
+				User:     "user:github|alice@openfga.com",
 			}}},
 		},
-		allowSchema10: true,
-
 		// output
-		err: serverErrors.WriteToIndirectRelationError("Attempting to write directly to an indirect only relationship", &openfgapb.TupleKey{
-			Object:   "repository:openfga/openfga",
-			Relation: "viewer",
-			User:     "github|alice@openfga.com",
-		}),
+		err: serverErrors.ValidationError(
+			&tuple.InvalidTupleError{
+				Cause:    fmt.Errorf("type 'user' is not an allowed type restriction for 'repo#viewer'"),
```

Comment:
I'm looking at https://app.codecov.io/gh/openfga/openfga/pull/763/indirect-changes and we aren't hitting this code path:  https://github.com/openfga/openfga/blob/0f1a5d06073efe0fbd33823cda56a6202864f816/pkg/server/commands/write.go#L102-L104 anymore.

I feel like `"Attempting to write directly to an indirect only relationship"` is clearer than `"type 'user' is not an allowed type restriction for 'repo#viewer'"` . `repo#viewer` doesn't support writing, doesn't matter which type you try.

Any chance of bringing back the previous error result?

Reply:
I go both ways, but I see where you're coming from.

To me the type restrictions are the set of allowed types that the relation can be related to, and a set of allowed types that is greater than 1 implies the relationship is, in some way, related to itself (e.g. `as self` shows up somewhere in the rewrite rule) and therefore is assignable.

If there are no type restrictions specified (e.g. `directly_related_user_types: []`), then that implies there are no types allowed to be written with that relation and therefore is unassignable. We can call this an indirect relationship or a purely rewritten relationship (I like to call it an unasignable relation), but I think for the sake of uniformity and to reduce the variation of errors it makes sense to leave it as is and remove this code bit.

Maybe a change of the error message itself would help?

Go from
> type 'user' is not an allowed type restriction for 'repo#viewer'

to
> object's of type 'user' are not an allowed relationship with the relation 'repo#viewer'

Reply:
Your suggestion still has the issue that me as a consumer, if i get the error `object's of type 'user' are not an allowed relationship with the relation 'repo#viewer'` then i will tend to think "what if I try writing an object of type "group" or "team"?... but i'll get a similar error. 

`"Attempting to write directly to an indirect only relationship"` implies that _no_ type is allowed

Reply:
`group` or `team` are not valid type restrictions just like `user` is not. If it doesn't show up in the type restriction list (or if the list is empty), then are no valid relationships that are valid targets to be written.

Reply:
let's resurrect in a future PR if we want to